common/src/main/java/com/hortonworks/registries/common/util/HdfsFileStorage.java

```diff
@@ -33,68 +44,155 @@
+        Preconditions.checkArgument(fsUrl != null, "fsUrl must be specified for HdfsFileStorage.");
+
+        Preconditions.checkArgument((keytabLocation == null) == (kerberosPrincipal == null),
+                "Both %s (== %s) and %s (== %s) must be specified to use kerberos.",
+                CONFIG_KERBEROS_PRINCIPAL, kerberosPrincipal, CONFIG_KERBEROS_KEYTAB, keytabLocation);
+
+        if (kerberosPrincipal != null) {
+            LOG.info("Logging in as kerberos principal {}", kerberosPrincipal);
+            UserGroupInformation.loginUserFromKeytab(kerberosPrincipal, keytabLocation);
+            kerberos = true;
         }
 
-        try {
-            hdfsFileSystem = FileSystem.get(URI.create(fsUrl), hdfsConfig);
-        } catch (IOException e) {
-            throw new RuntimeException(e);
+        directory = adjustDirectory(fsUrl, directory);
+        fsUri = URI.create(fsUrl);
+
+        LOG.info("Initialized with fsUrl={}, directory={}, kerberos principal={}", fsUrl, directory, kerberosPrincipal);
+    }
+
+    /**
+     * Cloud storage filesystem url's usally contain paths. These should be prepended to the directory in order to work
+     * properly.
+     * @param fsUrl The HDFS or compatible filesystem URL
+     * @param directory the directory to store hars
+     * @return directory adjusted with path component in fsUrl
+     */
+    @VisibleForTesting
+    static String adjustDirectory(String fsUrl, String directory) {
+        String pathInFsUrl = URI.create(fsUrl).getPath();
+        if ( !(pathInFsUrl.endsWith("/") || directory.startsWith("/")) ) {
+            return pathInFsUrl + "/" + directory;
+        }
+        else if (pathInFsUrl.endsWith("/") && directory.startsWith("/")) {
+            return pathInFsUrl + directory.substring(1);
+        }
+        else {
+            return pathInFsUrl + directory;
         }
     }
 
+    private boolean isKerberos() {
+        return kerberos;
+    }
+
     @Override
     public String upload(InputStream inputStream, String name) throws IOException {
-        Path jarPath = new Path(directory, name);
-
-        try(FSDataOutputStream outputStream = hdfsFileSystem.create(jarPath, false)) {
-            ByteStreams.copy(inputStream, outputStream);
-        }
-
-        return jarPath.toString();
+        return isKerberos() ?
+                doAsAndConvertException(() -> uploadInternal(inputStream, name)):
+                uploadInternal(inputStream, name);
     }
 
     @Override
     public InputStream download(String name) throws IOException {
-        Path filePath = new Path(directory, name);
-        return hdfsFileSystem.open(filePath);
+        return isKerberos() ?
+                doAsAndConvertException(() -> downloadInternal(name)):
+                downloadInternal(name);
     }
 
     @Override
     public boolean delete(String name) throws IOException {
-        return hdfsFileSystem.delete(new Path(directory, name), true);
+        return isKerberos() ?
```

Comment:
upload(), download() and delete() all have similar structure. We check if kerberos is enabled then we call doAsAndConvertException(()->internalMethod()) or internalMethod(). We can refactor this.